### ISSUE TITLE
Internal Service Errors after deploying Balrog

### ISSUE THREAD

**Issue by ahal** [mozilla-releng/balrog#2479]
SRE Services deployed Balrog around 17:00 EDT. At 17:45 EDT we started seeing balrogscript task failures and the queue grew to 1000+ tasks. Sheriffs alerted #firefox-ci and #taskcluster-cloudops at 21:00 EDT. Balrog web was also hitting network errors on all pages.

JBuck rolled back the deploy and everything cleared up immediately. He was able to grab some stack traces first:
```
"Uncaught exception:
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 2073, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1519, in full_dispatch_request
    return self.finalize_request(rv)
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1540, in finalize_request
    response = self.process_response(response)
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1885, in process_response
    response = self.ensure_sync(func)(response)
  File "/usr/local/lib/python3.9/site-packages/balrog-3.15-py3.9.egg/auslib/web/admin/base.py", line 93, in complete_request
    request.transaction.rollback()
<class 'AttributeError'>
AttributeError("'Request' object has no attribute 'transaction'")
"
```

and

```
"Uncaught exception:
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1540, in finalize_request
    response = self.process_response(response)
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1885, in process_response
    response = self.ensure_sync(func)(response)
  File "/usr/local/lib/python3.9/site-packages/balrog-3.15-py3.9.egg/auslib/web/admin/base.py", line 93, in complete_request
    request.transaction.rollback()
<class 'AttributeError'>
AttributeError("'Request' object has no attribute 'transaction'")
"
```

Maybe something in the recent dependency updates / or upgrade to Python 3.9 is the cause here?

/cc @jcristau

**Comment by ahal** [comment]
See also https://bugzilla.mozilla.org/show_bug.cgi?id=1761398

**Comment by ahal** [comment]
The exception appears to be happening in a failure path, so it's possibly hiding the true problem.

**Comment by jbuck** [comment]
We deployed v3.15 today, and rolling back the admin to v3.14 fixed things up

**Comment by gbrownmozilla** [comment]
We have seen similar AttributeErrors before (Jan 9, 2022):
https://sentry.prod.mozaws.net/operations/prod-admin/issues/16816968/?query=is%3Aunresolved

**Comment by gbrownmozilla** [comment]
A typical failed request was
https://aus4-admin.mozilla.org/api/v2/releases/Firefox-99.0b8-build1

**Comment by ahal** [comment]
> We have seen similar AttributeErrors before (Jan 9, 2022): https://sentry.prod.mozaws.net/operations/prod-admin/issues/16816968/?query=is%3Aunresolved

Yeah, so I'm guessing the error handling is broken then.

**Comment by gbrownmozilla** [comment]
Earlier today, staging admin was still at 3.14. I tried hitting it with some moderate load:
```
$ curl --parallel --parallel-immediate --parallel-max 20 --config urls > oo14
DL% UL%  Dled  Uled  Xfers  Live   Qd Total     Current  Left    Speed
--  --  2434k     0   114     0    47 --:--:--  0:00:09 --:--:--  273k      
```
(urls contains various staging admin urls like "url = https://admin-stage.balrog.nonprod.cloudops.mozgcp.net/api/v2/releases/Firefox-99.0b9-build1")
I didn't see any new errors in Sentry.

I nudged SRE to deploy 3.15 on staging admin, and repeated:
```
$ curl --parallel --parallel-immediate --parallel-max 20 --config urls > oo15
DL% UL%  Dled  Uled  Xfers  Live   Qd Total     Current  Left    Speed
--  --  2434k     0   114     0    47 --:--:--  0:00:34 --:--:-- 35815  
```
It ran much slower (but maybe that's coincidence / temporary) and did produce some ISE errors (paired with AttributeErrors) in Sentry -- but staging continued to respond to all queries.

The production outage seemed to involve POST requests, so these curl GETs may not be the best test, but might be telling us something?

**Comment by gbrownmozilla** [comment]
In addition to the ISE and AttributeErrors, my staging admin tests also produced some TimeoutErrors:
```
TimeoutError

QueuePool limit of size 5 overflow 10 reached, connection timed out, timeout 30.00 (Background on this error at: https://sqlalche.me/e/14/3o7r)
```
but I don't see those on production, so maybe this is unrelated.

**Comment by gbrownmozilla** [comment]
I'm not sure I understand the pairs of (ISE, AttributeError): Looking at an AttributeError in Sentry, like https://sentry.io/organizations/mozilla/issues/3129377324/?project=6262508, I see the AttributeError happens at https://github.com/mozilla-releng/balrog/blob/main/src/auslib/web/admin/base.py#L93, but its "breadcrumbs" include "Caught ISE 500 errors". Is the AttributeError reported to Sentry at https://github.com/mozilla-releng/balrog/blob/main/src/auslib/web/admin/base.py#L136? That is, does the AttributeError "cause" the ISE? If that's the case, what causes the AttributeError -- why is  `response.status_code >= 400 ` in complete_response()? Or do I have it all backwards?

**Comment by gbrownmozilla** [comment]
Today, my concurrent requests test on staging admin is even slower:
```
$ curl --parallel --parallel-immediate --parallel-max 20 --config urls > oo15
DL% UL%  Dled  Uled  Xfers  Live   Qd Total     Current  Left    Speed
--  --  2417k     0   114     0    46 --:--:--  0:01:04 --:--:--    39      
```
If I run Balrog locally and test against that, the response is much faster (and constant - it never seems to degrade), even if I retrieve more urls:
```
$ curl --parallel --parallel-immediate --parallel-max 20 --insecure --config localurls > oolocal15
DL% UL%  Dled  Uled  Xfers  Live   Qd Total     Current  Left    Speed
--  --  22.4M     0   129     0    55 --:--:--  0:00:06 --:--:-- 5155k      
```
(notice that is 10x as much data in 10x less time).

But how much of that difference is spent on the network?

**Comment by gbrownmozilla** [comment]
Also, running locally I don't see those TimeoutErrors or ISEs or AttributeErrors. But, I'm noticing those on staging because of Sentry; when I run locally, am I just missing them when I manually scan the logs?

In short, I cannot reproduce any trouble with Balrog 3.15 locally, but it's also unclear if my local testing is effective.

**Comment by gbrownmozilla** [comment]
I started looking at gcloud logs for production during the outage. The 500 series failures are evident, but I don't see anything else out of the ordinary, so far.
https://console.cloud.google.com/logs/query;timeRange=2022-03-24T20:30:14.115Z%2F2022-03-24T23:41:14.115Z;cursorTimestamp=2022-03-24T23:37:35.641994851Z?project=moz-fx-balrog-prod-3fa2&pli=1&rapt=AEjHL4MIZZS3EDX8cZo12DsdgXnGOMYvOEgVQWEnyFBI8RcR10KAbZrt_BLF1YYlc6DDAJWD9Bw54m20aDd66NCq0LvF0CGjlA

**Comment by jcristau** [comment]
#2497 may fix this.  I was able to reproduce the outage on staging today, so I'm planning on tagging 3.16 and testing again on stage next week.

**Comment by gbrownmozilla** [comment]
With staging broken now (thanks Julien!) I see __version__ is still responsive, but __heartbeat__ is not -- more evidence that the database is central to the cause.

**Comment by jcristau** [comment]
3.16 tagged and deployed to staging.  The same steps that broke it last week (i.e. a staging beta release with all locales enabled) seem to have gone just fine today, and balrog-admin-stage remains responsive.

Closing as fixed by #2497.
